pkg/controller/warehouses/warehouses.go

```diff
@@ -284,6 +288,74 @@ func (r *reconciler) syncWarehouse(
 	// if the Warehouse is configured to do so.
 	if pol := warehouse.Spec.FreightCreationPolicy; pol == kargoapi.FreightCreationPolicyAutomatic || pol == "" {
+		criteriaSatisfied, err := freightCreationCriteriaSatisfied(ctx,
+			warehouse.Spec.FreightCreationCriteria,
+			status.DiscoveredArtifacts,
+		)
+		if err != nil {
+			logger.Error(err, "error evaluating freight creation criteria")
+			conditions.Set(
+				&status,
+				&metav1.Condition{
+					Type:               kargoapi.ConditionTypeHealthy,
+					Status:             metav1.ConditionFalse,
+					Reason:             "CriteriaEvaluationFailed",
+					Message:            fmt.Sprintf("Criteria evaluation failed: %s", err.Error()),
+					ObservedGeneration: warehouse.GetGeneration(),
+				},
+				&metav1.Condition{
+					Type:               kargoapi.ConditionTypeReady,
+					Status:             metav1.ConditionFalse,
+					Reason:             "CriteriaEvaluationFailed",
+					Message:            fmt.Sprintf("Criteria evaluation failed: %s", err.Error()),
+					ObservedGeneration: warehouse.GetGeneration(),
+				},
+				&metav1.Condition{
+					Type:               kargoapi.ConditionTypeCriteriaSatisfied,
+					Status:             metav1.ConditionUnknown,
+					Reason:             "FreightCreationCriteriaError",
+					Message:            fmt.Sprintf("Criteria evaluation failed: %s", err.Error()),
+					ObservedGeneration: warehouse.GetGeneration(),
+				},
+			)
+			conditions.Delete(&status, kargoapi.ConditionTypeReconciling)
+			// return a nil error to avoid a requeue loop since subsequent
+			// retries are not going to make the expression any more valid.
+			return status, nil
+		}
+		if !criteriaSatisfied {
+			logger.Debug("freight creation criteria not satisfied; skipping freight creation")
+			conditions.Set(
+				&status,
+				&metav1.Condition{
+					Type:               kargoapi.ConditionTypeHealthy,
+					Status:             metav1.ConditionTrue,
+					Reason:             "ArtifactsDiscovered",
+					Message:            conditions.Get(&status, kargoapi.ConditionTypeHealthy).Message,
+					ObservedGeneration: warehouse.GetGeneration(),
+				},
+				&metav1.Condition{
+					Type:   kargoapi.ConditionTypeCriteriaSatisfied,
+					Status: metav1.ConditionFalse,
+					Reason: "InsufficientChanges",
+					Message: fmt.Sprintf(
+						"None of the discovered artifacts could match the Freight creation criteria %q",
+						warehouse.Spec.FreightCreationCriteria.Expression,
+					),
+					ObservedGeneration: warehouse.GetGeneration(),
+				},
+				&metav1.Condition{
+					Type:               kargoapi.ConditionTypeReady,
+					Status:             metav1.ConditionTrue,
+					Reason:             "ArtifactsDiscovered",
+					Message:            conditions.Get(&status, kargoapi.ConditionTypeHealthy).Message,
+					ObservedGeneration: warehouse.GetGeneration(),
+				},
+			)
+			conditions.Delete(&status, kargoapi.ConditionTypeReconciling)
+			return status, nil
```

Comment:
I think there's more stuff happening here than is necessary.

I would consider not updating Healthy or Ready conditions here and not deleting the reconciling conditions...

I would set the CriteriaSatisfied condition and then _not_ return early...

I'd put an `else` after this that isn't currently there, handle Freight creation within that block and then, only after _both_ the existing `if` and the new `else`, tidy up all the conditions to reflect success.

imho, all of this code becomes easier to follow if all successful paths through the code lead to conditions being set consistently in one place.

cc @hiddeco